VR/VRRenderer.cpp

```diff
@@ -0,0 +1,449 @@
+    waitFrameInfo.type = XR_TYPE_FRAME_WAIT_INFO;
+    waitFrameInfo.next = NULL;
+
+    frameState.type = XR_TYPE_FRAME_STATE;
+    frameState.next = NULL;
+
+    OXR(xrWaitFrame(engine->appState.Session, &waitFrameInfo, &frameState));
+    engine->predictedDisplayTime = frameState.predictedDisplayTime;
+    if (!frameState.shouldRender) {
+        return;
+    }
+
+    // Get the HMD pose, predicted for the middle of the time period during which
+    // the new eye images will be displayed. The number of frames predicted ahead
+    // depends on the pipeline depth of the engine and the synthesis rate.
+    // The better the prediction, the less black will be pulled in at the edges.
+    XrFrameBeginInfo beginFrameDesc = {};
+    beginFrameDesc.type = XR_TYPE_FRAME_BEGIN_INFO;
+    beginFrameDesc.next = NULL;
+    OXR(xrBeginFrame(engine->appState.Session, &beginFrameDesc));
+
+    XrViewLocateInfo projectionInfo = {};
+    projectionInfo.type = XR_TYPE_VIEW_LOCATE_INFO;
+    projectionInfo.viewConfigurationType = engine->appState.ViewportConfig.viewConfigurationType;
+    projectionInfo.displayTime = frameState.predictedDisplayTime;
+    projectionInfo.space = engine->appState.CurrentSpace;
+
+    XrViewState viewState = {XR_TYPE_VIEW_STATE, NULL};
+
+    uint32_t projectionCapacityInput = ovrMaxNumEyes;
+    uint32_t projectionCountOutput = projectionCapacityInput;
+
+    OXR(xrLocateViews(
+            engine->appState.Session,
+            &projectionInfo,
+            &viewState,
+            projectionCapacityInput,
+            &projectionCountOutput,
+            projections));
+    //
+
+    for (int eye = 0; eye < ovrMaxNumEyes; eye++) {
+        invViewTransform[eye] = projections[eye].pose;
+
+        fov.angleLeft += projections[eye].fov.angleLeft / 2.0f;
+        fov.angleRight += projections[eye].fov.angleRight / 2.0f;
+        fov.angleUp += projections[eye].fov.angleUp / 2.0f;
+        fov.angleDown += projections[eye].fov.angleDown / 2.0f;
+    }
+
+    // Update HMD and controllers
+    vec3_t rotation = {0, 0, 0};
+    QuatToYawPitchRoll(invViewTransform[0].orientation, rotation, hmdorientation);
+	hmdposition[0] = invViewTransform[0].position.x;
+	hmdposition[1] = invViewTransform[0].position.y;
+	hmdposition[2] = invViewTransform[0].position.z;
+	IN_VRInputFrame(engine);
+
+    const ovrMatrix4f projectionMatrix = ovrMatrix4f_CreateProjectionFov(
+            fov.angleLeft, fov.angleRight, fov.angleUp, fov.angleDown, 1.0f, 0.0f );
+
+    engine->appState.LayerCount = 0;
+    memset(engine->appState.Layers, 0, sizeof(ovrCompositorLayer_Union) * ovrMaxLayerCount);
+
+	//TODO:
+    /*re.SetVRHeadsetParms(projectionMatrix.M,
+                         engine->appState.Renderer.FrameBuffer[0].FrameBuffers[engine->appState.Renderer.FrameBuffer[0].TextureSwapChainIndex],
+                         engine->appState.Renderer.FrameBuffer[1].FrameBuffers[engine->appState.Renderer.FrameBuffer[1].TextureSwapChainIndex]);*/
```

Comment:
Expand on this TODO: comment?

Reply:
This code is no longer needed, the VR matrices are accessible using VR_GetMatrix, the framebuffers using VR_BindFramebuffer.